### ISSUE TITLE
Tests failed

### ISSUE THREAD

**Issue by balasankarc** [flavorjones/loofah#87]
When I did a `bundle exec rake` , I got the following error

    # Running tests:

    ............................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................F.........................................F.........F.......................F...............SS..................

    Finished tests in 0.715461s, 1134.9321 tests/s, 1287.2814 assertions/s.

      1) Failure:
    test_fragment_whitewash_on_microsofty_markup(IntegrationTestAdHoc) [/home/balasankarc/packaging/loofah/source/loofah-2.0.1/test/integration/test_ad_hoc.rb:146]:
    --- expected
    +++ actual
    @@ -1 +1,4 @@
    -"<p>Foo <b>BOLD</b></p>"
    +"
    +
    +<p>Foo <b>BOLD</b></p>
    +"

      2) Failure:
    test_fragment_whitewash_on_microsofty_markup(blank input string) [/home/balasankarc/packaging/loofah/source/loofah-2.0.1/test/integration/test_ad_hoc.rb:146]:
    --- expected
    +++ actual
    @@ -1 +1,4 @@
    -"<p>Foo <b>BOLD</b></p>"
    +"
    +
    +<p>Foo <b>BOLD</b></p>
    +"

      3) Failure:
    test_fragment_whitewash_on_microsofty_markup(blank input string::document) [/home/balasankarc/packaging/loofah/source/loofah-2.0.1/test/integration/test_ad_hoc.rb:146]:
    --- expected
    +++ actual
    @@ -1 +1,4 @@
    -"<p>Foo <b>BOLD</b></p>"
    +"
    +
    +<p>Foo <b>BOLD</b></p>
    +"

      4) Failure:
    test_fragment_whitewash_on_microsofty_markup(blank input string::fragment) [/home/balasankarc/packaging/loofah/source/loofah-2.0.1/test/integration/test_ad_hoc.rb:146]:
    --- expected
    +++ actual
    @@ -1 +1,4 @@
    -"<p>Foo <b>BOLD</b></p>"
    +"
    +
    +<p>Foo <b>BOLD</b></p>
    +"

    812 tests, 921 assertions, 4 failures, 0 errors, 2 skips
    rake abort

Can you tell me what the issue is?

PS : I am packaging loofah for Debian and if these tests could be ran succesfully, it would be a great help.

**Comment by flavorjones** [comment]
Hi,

Thanks for asking this question, and working towards packaging Loofah.

I'll need more information in order to reproduce what you're seeing. Can you please include the *full* output from your test suite? There's a line that begins "testing with Nokogiri ..." that contains lots of valuable information about your system.

For example, on my system I see:

```
=> testing with Nokogiri {"warnings"=>[], "nokogiri"=>"1.6.6.2", "ruby"=>{"version"=>"2.2.2", "platform"=>"x86_64-linux", "description"=>"ruby 2.2.2p95 (2015-04-13 revision 50295) [x86_64-linux]", "engine"=>"ruby"}, "libxml"=>{"binding"=>"extension", "source"=>"packaged", "libxml2_path"=>"/home/flavorjones/.rvm/gems/ruby-2.2.2/gems/nokogiri-1.6.6.2/ports/x86_64-unknown-linux-gnu/libxml2/2.9.2", "libxslt_path"=>"/home/flavorjones/.rvm/gems/ruby-2.2.2/gems/nokogiri-1.6.6.2/ports/x86_64-unknown-linux-gnu/libxslt/1.1.28", "libxml2_patches"=>["0001-Revert-Missing-initialization-for-the-catalog-module.patch", "0002-Fix-missing-entities-after-CVE-2014-3660-fix.patch"], "libxslt_patches"=>["0001-Adding-doc-update-related-to-1.1.28.patch", "0002-Fix-a-couple-of-places-where-f-printf-parameters-wer.patch", "0003-Initialize-pseudo-random-number-generator-with-curre.patch", "0004-EXSLT-function-str-replace-is-broken-as-is.patch", "0006-Fix-str-padding-to-work-with-UTF-8-strings.patch", "0007-Separate-function-for-predicate-matching-in-patterns.patch", "0008-Fix-direct-pattern-matching.patch", "0009-Fix-certain-patterns-with-predicates.patch", "0010-Fix-handling-of-UTF-8-strings-in-EXSLT-crypto-module.patch", "0013-Memory-leak-in-xsltCompileIdKeyPattern-error-path.patch", "0014-Fix-for-bug-436589.patch", "0015-Fix-mkdir-for-mingw.patch"], "compiled"=>"2.9.2", "loaded"=>"2.9.2"}}
Run options: --seed 53560

# Running tests:

......................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................SS....................

Finished tests in 0.308359s, 2633.2924 tests/s, 2986.7762 assertions/s.

812 tests, 921 assertions, 0 failures, 0 errors, 2 skips
```

The kind of failures you're seeing (essentially, whitespace differences) tends to be very sensitive to the version of libxml2 you're running. As you can see above, I'm running libxml2 2.9.2. You are likely running libxml2 2.9.1, which appears to be the default in Jessie.

Unfortunately, if this is the case, you may not get a satisfactory answer from me. We found v2.9.1 to be problematic, particularly around [this XPath bug](https://bugzilla.gnome.org/show_bug.cgi?id=695699), not fixed until v2.9.2 which is why Nokogiri by default tries hard to install that version.

In other words, Nokogiri's test suite won't pass on libxml2.9.1, and as a result Loofah will likely exhibit buggy behavior as a result (at least if CSS or XPath queries involving node position are used).

Thanks,
-mike

**Comment by balasankarc** [comment]
I am on Debian Sid and am using libxml2 v2.9.2. This is the full output trace.

$ bundle exec rake

    /home/balasankarc/.rvm/rubies/ruby-2.0.0-p643/bin/ruby -w -Ilib:bin:test:. -e 'require "rubygems"; require "minitest/autorun"; require "test/unit/test_helpers.rb"; require "test/unit/test_encoding.rb"; require "test/unit/test_scrubber.rb"; require "test/unit/test_api.rb"; require "test/unit/test_scrubbers.rb"; require "test/html5/test_sanitizer.rb"; require "test/integration/test_xml.rb"; require "test/integration/test_helpers.rb"; require "test/integration/test_html.rb"; require "test/integration/test_scrubbers.rb"; require "test/integration/test_ad_hoc.rb"' -- 
    => testing with Nokogiri {"warnings"=>[], "nokogiri"=>"1.6.6.2", "ruby"=>{"version"=>"2.0.0", "platform"=>"i686-linux", "description"=>"ruby 2.0.0p643 (2015-02-25 revision 49749) [i686-linux]", "engine"=>"ruby"}, "libxml"=>{"binding"=>"extension", "source"=>"packaged", "libxml2_path"=>"/home/balasankarc/.rvm/gems/ruby-2.0.0-p643/gems/nokogiri-1.6.6.2/ports/i686-pc-linux-gnu/libxml2/2.9.2", "libxslt_path"=>"/home/balasankarc/.rvm/gems/ruby-2.0.0-p643/gems/nokogiri-1.6.6.2/ports/i686-pc-linux-gnu/libxslt/1.1.28", "libxml2_patches"=>["0001-Revert-Missing-initialization-for-the-catalog-module.patch", "0002-Fix-missing-entities-after-CVE-2014-3660-fix.patch"], "libxslt_patches"=>["0001-Adding-doc-update-related-to-1.1.28.patch", "0002-Fix-a-couple-of-places-where-f-printf-parameters-wer.patch", "0003-Initialize-pseudo-random-number-generator-with-curre.patch", "0004-EXSLT-function-str-replace-is-broken-as-is.patch", "0006-Fix-str-padding-to-work-with-UTF-8-strings.patch", "0007-Separate-function-for-predicate-matching-in-patterns.patch", "0008-Fix-direct-pattern-matching.patch", "0009-Fix-certain-patterns-with-predicates.patch", "0010-Fix-handling-of-UTF-8-strings-in-EXSLT-crypto-module.patch", "0013-Memory-leak-in-xsltCompileIdKeyPattern-error-path.patch", "0014-Fix-for-bug-436589.patch", "0015-Fix-mkdir-for-mingw.patch"], "compiled"=>"2.9.2", "loaded"=>"2.9.2"}}
    Run options: --seed 16415

    # Running tests:

    .............................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................F..........................................F.....F.......................F.................SS..................

    Finished tests in 0.748346s, 1085.0591 tests/s, 1230.7136 assertions/s.

      1) Failure:
    test_fragment_whitewash_on_microsofty_markup(IntegrationTestAdHoc) [/home/balasankarc/packaging/loofah/source/loofah-2.0.1/test/integration/test_ad_hoc.rb:146]:
    --- expected
    +++ actual
    @@ -1 +1,4 @@
    -"<p>Foo <b>BOLD</b></p>"
    +"
    +
    +<p>Foo <b>BOLD</b></p>
    +"


      2) Failure:
    test_fragment_whitewash_on_microsofty_markup(blank input string) [/home/balasankarc/packaging/loofah/source/loofah-2.0.1/test/integration/test_ad_hoc.rb:146]:
    --- expected
    +++ actual
    @@ -1 +1,4 @@
    -"<p>Foo <b>BOLD</b></p>"
    +"
    +
    +<p>Foo <b>BOLD</b></p>
    +"


      3) Failure:
    test_fragment_whitewash_on_microsofty_markup(blank input string::document) [/home/balasankarc/packaging/loofah/source/loofah-2.0.1/test/integration/test_ad_hoc.rb:146]:
    --- expected
    +++ actual
    @@ -1 +1,4 @@
    -"<p>Foo <b>BOLD</b></p>"
    +"
    +
    +<p>Foo <b>BOLD</b></p>
    +"


      4) Failure:
    test_fragment_whitewash_on_microsofty_markup(blank input string::fragment) [/home/balasankarc/packaging/loofah/source/loofah-2.0.1/test/integration/test_ad_hoc.rb:146]:
    --- expected
    +++ actual
    @@ -1 +1,4 @@
    -"<p>Foo <b>BOLD</b></p>"
    +"
    +
    +<p>Foo <b>BOLD</b></p>
    +"


    812 tests, 921 assertions, 4 failures, 0 errors, 2 skips
    rake aborted!
    Command failed with status (1): [/home/balasankarc/.rvm/rubies/ruby-2.0.0-p...]
    /home/balasankarc/.rvm/gems/ruby-2.0.0-p643/gems/hoe-3.13.1/lib/hoe/test.rb:83:in `block in define_test_tasks'
    /home/balasankarc/.rvm/gems/ruby-2.0.0-p643/bin/ruby_executable_hooks:15:in `eval'
    /home/balasankarc/.rvm/gems/ruby-2.0.0-p643/bin/ruby_executable_hooks:15:in `<main>'
    Tasks: TOP => default => test
    (See full trace by running task with --trace)

**Comment by balasankarc** [comment]
Ah.. I was trying to package the last release of loofah - 2.0.1 and it was giving test failures. But when I cloned the repo now and try rake, tests ran successfully. 

Is the current state of git repo valid for a new release? If yes, can please do so. Thanks.

**Comment by flavorjones** [comment]
Hi, cutting v2.0.2 this afternoon.

**Comment by balasankarc** [comment]
Thanks. :)

**Comment by flavorjones** [comment]
In case you didn't see it, v2.0.2. is up at https://rubygems.org/gems/loofah

**Comment by balasankarc** [comment]
Yup. I saw it.. I've completed the packaging and submitted it to Debian. It is currently in New[1] queue and hopefully will get accepted soon.

[1] https://ftp-master.debian.org/new.html

**Comment by balasankarc** [comment]
It is accepted to Debian. Thanks for your help.

**Comment by flavorjones** [comment]
Excellent news!
